snapshotter/app/service.go

```diff
@@ -166,58 +168,51 @@ func initSnapshotter(ctx context.Context, config config.Config, cache cache.Cach
-		// server lifecycle. tracked here: https://github.com/firecracker-microvm/firecracker-containerd/issues/607
-		portMap := make(map[int]bool)
-		if config.Snapshotter.Metrics.Enable {
-			metricsPort, err := strconv.ParseUint(response.MetricsPort, base10, bits32)
-			if err != nil {
-				return nil, err
-			}
+		metricsProxy, err := initMetricsProxy(config, monitor, host, response.MetricsPort, response.Labels)
+		if err != nil {
+			return nil, err
+		}
 
-			metricsDialer := func(ctx context.Context, _, _ string) (net.Conn, error) {
-				return vsock.DialContext(ctx, host, uint32(metricsPort), vsock.WithLogger(log.G(ctx)))
-			}
+		return proxy.NewRemoteSnapshotter(ctx, host, snapshotterDialer, metricsProxy)
+	}
+
+	return demux.NewSnapshotter(cache, newRemoteSnapshotterFunc), nil
+}
 
-			portRange := config.Snapshotter.Metrics.PortRange
-			metricsHost := config.Snapshotter.Metrics.Host
+func initMetricsProxyMonitor(portRange string) (*metrics.Monitor, error) {
+	ports := strings.Split(portRange, "-")
+	portRangeError := fmt.Errorf("invalid port range %s", portRange)
+	if len(ports) < 2 {
+		return nil, portRangeError
+	}
+	lower, err := strconv.Atoi(ports[0])
+	if err != nil {
+		return nil, portRangeError
+	}
+	upper, err := strconv.Atoi(ports[1])
+	if err != nil {
+		return nil, portRangeError
+	}
 
-			// Assign a port for metrics proxy server.
-			ports := strings.Split(portRange, "-")
-			portRangeError := fmt.Errorf("invalid port range %s", portRange)
-			if len(ports) < 2 {
-				return nil, portRangeError
-			}
-			lower, err := strconv.Atoi(ports[0])
-			if err != nil {
-				return nil, portRangeError
-			}
-			upper, err := strconv.Atoi(ports[1])
-			if err != nil {
-				return nil, portRangeError
-			}
-			port := -1
-			for p := lower; p <= upper; p++ {
-				if _, ok := portMap[p]; !ok {
-					port = p
-					portMap[p] = true
-					break
-				}
-			}
-			if port < 0 {
-				return nil, fmt.Errorf("invalid port: %d", port)
-			}
+	return metrics.NewMonitor(lower, upper), nil
+}
 
-			metricsProxy, err = metrics.NewProxy(metricsHost, port, response.Labels, metricsDialer)
-			if err != nil {
-				return nil, err
-			}
+func initMetricsProxy(config config.Config, monitor *metrics.Monitor, host, port string, labels map[string]string) (*metrics.Proxy, error) {
+	var metricsProxy *metrics.Proxy
+	metricsPort, err := strconv.ParseUint(port, base10, bits32)
+	if err != nil {
+		return nil, err
+	}
 
-		}
+	metricsDialer := func(ctx context.Context, _, _ string) (net.Conn, error) {
+		return vsock.DialContext(ctx, host, uint32(metricsPort), vsock.WithLogger(log.G(ctx)))
+	}
 
-		return proxy.NewProxySnapshotter(ctx, host, snapshotterDialer, metricsProxy)
+	metricsHost := config.Snapshotter.Metrics.Host
+	metricsProxy, err = metrics.NewProxy(metricsHost, monitor, labels, metricsDialer)
+	if err != nil {
+		return nil, err
 	}
 
-	return demux.NewSnapshotter(cache, newProxySnapshotterFunc), nil
+	return metricsProxy, nil
```

Comment:
No special error handling here right now. Could be more concise.

```suggestion
return metrics.NewProxy(metricsHost, monitor, labels, metricsDialer)
```

Reply:
gp, I'll make this change 👍